### PR TITLE
adopt mlx-swift 0.25.2

### PR DESCRIPTION
**Pull request by davidkoski**
Changes for adopting mlx / mlx-swift 0.25.2

- #274 
- #273

### CODE REVIEW

File: Libraries/MLXLMCommon/KVCache.swift [hunk 1]
```diff
@@ -33,11 +43,38 @@ public func createAttentionMask(h: MLXArray, cache: [KVCache]?) -> MLXArray? {
...
+        }
+
+        if returnArray {
+            return .array(createAdditiveCausalMask(n: t, offset: offset, windowSize: windowSize))
+        } else {
+            return .causal
+        }
+    }
+    return .none
+}
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/KVCache.swift:76]:
```python

def create_attention_mask(
    h: mx.array, cache: Optional[Any] = None, return_array: bool = False
):
    T = h.shape[1]
    if T > 1:
        offset = 0
        window_size = None
        if cache is not None and cache[0] is not None:
            c = cache[0]
            offset = c.offset
            if hasattr(c, "max_size"):
                window_size = c.max_size
                offset = min(window_size, offset)
                return_array = return_array or offset + T > window_size
        if return_array:
            return create_causal_mask(T, offset, window_size=window_size)
        else:
            return "causal"
    else:
        mask = None
    return mask
```

File: Libraries/MLXLMCommon/KVCache.swift [hunk 2]
```diff
@@ -12,19 +12,29 @@ public protocol KVCache: Evaluatable {
...
+    var linds = offset != 0 ? MLXArray(Int32(offset) ..< Int32(offset + n)) : rinds
+    linds = linds[0..., .newAxis]
+    rinds = rinds[.newAxis]
+    var mask = linds .>= rinds
+
+    if let windowSize {
+        mask = mask & (linds .< rinds + windowSize)
+    }
+
+    return mask
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/KVCache.swift:30]:
```python
def create_causal_mask(
    N: int,
    offset: int = 0,
    window_size: Optional[int] = None,
    lengths: Optional[mx.array] = None,
):
    rinds = mx.arange(offset + N)
    linds = mx.arange(offset, offset + N) if offset else rinds
    linds = linds[:, None]
    rinds = rinds[None]
    mask = linds >= rinds
    if window_size is not None:
        mask = mask & (linds <= rinds + window_size)
    if lengths is not None:
        lengths = lengths[:, None, None, None]
        mask = mask & (rinds < lengths)
    return mask
```

File: Applications/LLMEval/ContentView.swift [hunk 3]
```diff
@@ -4,7 +4,6 @@ import AsyncAlgorithms
 import MLX
 import MLXLLM
 import MLXLMCommon
-import MLXRandom
```
Comment (davidkoski) [inline Applications/LLMEval/ContentView.swift:7]: MLXRandom is folded into MLX now -- not required, but clean up the imports

File: Libraries/Embedders/Bert.swift [hunk 4]
```diff
@@ -41,7 +40,7 @@ private class BertEmbedding: Module {
         tokenTypeIds: MLXArray? = nil
     ) -> MLXArray {
         let posIds = positionIds ?? broadcast(MLXArray.arange(inputIds.dim(1)), to: inputIds.shape)
-        let words = wordEmbeddings(inputIds) + positionEmbeddings(posIds)
+        var words = wordEmbeddings(inputIds) + positionEmbeddings(posIds)
         if let tokenTypeIds, let tokenTypeEmbeddings {
             words += tokenTypeEmbeddings(tokenTypeIds)
```
Comment (davidkoski) [inline Libraries/Embedders/Bert.swift:45]: This should have been an error all along -- the `+=` should have required a `var`.  See https://github.com/ml-explore/mlx-swift/pull/193 -- due to an oddity there and the fact that `MLXArray` is a reference made this work.

File: Libraries/Embedders/Load.swift [hunk 5]
```diff
@@ -119,26 +118,3 @@ public func loadModelContainer(
     return try await ModelContainer(
         hub: hub, modelDirectory: modelDirectory, configuration: configuration)
 }
-
-// TODO remove once mlx-swift update is adopted
-func quantize(
```
Comment (davidkoski) [inline Libraries/Embedders/Load.swift:124]: Done

File: Libraries/Embedders/NomicBert.swift [hunk 6]
```diff
@@ -2,7 +2,6 @@
 
 import Foundation
 import MLX
-import MLXFast
```
Comment (davidkoski) [inline Libraries/Embedders/NomicBert.swift:5]: Also folded in to MLX, clean up the imports

File: Libraries/MLXLLM/LLMModelFactory.swift [hunk 7]
```diff
@@ -236,6 +236,7 @@ public class LLMRegistry: AbstractModelRegistry, @unchecked Sendable {
             qwen3_1_7b_4bit,
             qwen3_4b_4bit,
             qwen3_8b_4bit,
+            qwen3MoE_30b_a3b_4bit,
```
Comment (davidkoski) [inline Libraries/MLXLLM/LLMModelFactory.swift:239]: While testing the MoEs (see below) I noticed this was missing.

File: Libraries/MLXLLM/Models/Cohere.swift [hunk 8]
```diff
@@ -38,7 +37,7 @@ private class Attention: Module {
     }
 
     public func callAsFunction(
-        _ x: MLXArray, mask: MLXArray? = nil, cache: KVCache?
+        _ x: MLXArray, mask: MLXFast.ScaledDotProductAttentionMaskMode, cache: KVCache?
```
Comment (davidkoski) [inline Libraries/MLXLLM/Models/Cohere.swift:40]: Per #274 for fast SDPA / Attention code that don't do arithmetic on the MLXArray use the mode (which can carry an array or be `causal`).

File: Libraries/MLXLLM/Models/Cohere.swift [hunk 9]
```diff
@@ -136,7 +135,7 @@ public class CohereModelInner: Module {
     public func callAsFunction(_ inputs: MLXArray, cache: [KVCache]? = nil) -> MLXArray {
         var h = embedTokens(inputs)
 
-        let mask: MLXArray? = createAttentionMask(h: h, cache: cache)
+        let mask = createAttentionMask(h: h, cache: cache)
```
Comment (davidkoski) [inline Libraries/MLXLLM/Models/Cohere.swift:138]: #274 -- use this new variant to create the mode rather than a bare MLXArray

File: Libraries/MLXLLM/Models/Gemma2.swift [hunk 10]
```diff
@@ -61,7 +60,7 @@ private class Attention: Module {
     }
 
     public func callAsFunction(
-        _ x: MLXArray, mask: MLXArray? = nil, cache: KVCache?
+        _ x: MLXArray, mask: MLXArray?, cache: KVCache?
```
Comment (davidkoski) [inline Libraries/MLXLLM/Models/Gemma2.swift:63]: Instead of #274 we force MLXArray here as we do arithmetic below (matches python version):

```swift
        if let mask {
            scores = scores + mask
        }
```

File: Libraries/MLXLLM/SwitchLayers.swift [hunk 11]
```diff
@@ -1,11 +1,30 @@
...
+    )
+}
+
+private func scatterUnsort(x: MLXArray, invOrder: MLXArray, shape: [Int]? = nil) -> MLXArray {
+    var x = x[invOrder]
+    if let shape {
+        x = unflatten(x, axis: 0, shape: shape)
+    }
+    return x
+}
```
Comment (davidkoski) [inline Libraries/MLXLLM/SwitchLayers.swift:26]: Support for sorted `SwitchGLU`, #273

File: Libraries/MLXLLM/SwitchLayers.swift [hunk 12]
```diff
@@ -110,7 +147,7 @@ class SwitchLinear: Module, Quantizable {
     }
 }
 
-class QuantizedSwitchLinear: SwitchLinear {
+class QuantizedSwitchLinear: SwitchLinear, Quantized {
```
Comment (davidkoski) [inline Libraries/MLXLLM/SwitchLayers.swift:150]: Updates on the mlx-swift side require that this be marked as Quantized, otherwise as a subtype of Quantizable it will be detected as already quantized (and nothing will be done).

File: Libraries/MLXVLM/Models/Qwen25VL.swift [hunk 13]
```diff
@@ -550,14 +549,14 @@ private enum Vision {
             // Create attention mask
             let attentionMask = full(
                 [1, sequenceLength, sequenceLength],
-                values: Int8(-127))
+                values: false)
```
Comment (davidkoski) [inline Libraries/MLXVLM/Models/Qwen25VL.swift:552]: Switch to a bool mask (works with this version of mlx + mlx-swift)

**Review by awni — APPROVED** [review]
Looks great! Mostly straightforward diff.. let's land it!
